From 57175d37b146c088dc1ef5631d12b2ecb1117c6a Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Wed, 31 Jul 2024 11:52:05 +0200
Subject: [PATCH 1067/1695] drivers: remoteproc: stm32: Fix address size
 handling in DMA range parsing

This patch improves the DMA range parsing in the STM32 remote processor
driver by parsing the size of `dma-ranges`, taking into account 32-bit
and 64-bit CPU address formats.

The changes ensure that the DMA range parsing works correctly for
both 32-bit and 64-bit CPU architectures.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git c1e271fb997f93f044262c2e0778e884bcb1dce0 drivers: remoteproc: stm32: Fix address size handling in DMA range parsing]
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Change-Id: I0f3c928bd86986b8df8539921198306afb501b2f
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/396243
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/remoteproc/stm32_remoteproc.c | 32 ++++++++++++++++------
 1 file changed, 24 insertions(+), 8 deletions(-)

diff --git a/core/drivers/remoteproc/stm32_remoteproc.c b/core/drivers/remoteproc/stm32_remoteproc.c
index 77510cd08..05ae1556c 100644
--- a/core/drivers/remoteproc/stm32_remoteproc.c
+++ b/core/drivers/remoteproc/stm32_remoteproc.c
@@ -16,6 +16,7 @@
 #include <libfdt.h>
 #include <mm/core_memprot.h>
 #include <mm/core_mmu.h>
+#include <stdint.h>
 #if defined(CFG_STM32MP25) || defined(CFG_STM32MP23) || defined(CFG_STM32MP21)
 #include <stm32_sysconf.h>
 #endif
@@ -395,10 +396,15 @@ static TEE_Result stm32_rproc_get_dma_range(struct stm32_rproc_mem *region,
 					    const void *fdt, int node)
 {
 	const fdt32_t *list = NULL;
+	uint32_t size = 0;
 	int ahb_node = 0;
-	int len = 0;
+	int elt_size = 0;
+	uint32_t da = 0;
 	int nranges = 0;
+	paddr_t pa = 0;
+	int len = 0;
 	int i = 0;
+	int j = 0;
 
 	/*
 	 * The match between local and remote processor memory mapping is
@@ -416,15 +422,25 @@ static TEE_Result stm32_rproc_get_dma_range(struct stm32_rproc_mem *region,
 		return TEE_SUCCESS;
 	}
 
-	if ((len % (sizeof(uint32_t) * 3)))
-		return TEE_ERROR_GENERIC;
+	/* A dma-ranges element is constructed with:
+	 *  - the 32-bit remote processor address
+	 *  - the cpu address which depends on the CPU arch (32-bit or 64-bit)
+	 *  - the 32-bit remote processor memory mapping size
+	 */
+	elt_size = sizeof(uint32_t) + sizeof(paddr_t) + sizeof(uint32_t);
+
+	if (len % elt_size)
+		return TEE_ERROR_BAD_PARAMETERS;
 
-	nranges = len / sizeof(uint32_t);
+	nranges = len / elt_size;
 
-	for (i = 0; i < nranges; i += 3) {
-		uint32_t da = fdt32_to_cpu(list[i]);
-		uint32_t pa = fdt32_to_cpu(list[i + 1]);
-		uint32_t size = fdt32_to_cpu(list[i + 2]);
+	for (i = 0, j = 0; i < nranges; i++) {
+		da = fdt32_to_cpu(list[j++]);
+		pa = fdt32_to_cpu(list[j++]);
+#if defined(__LP64__)
+		pa = SHIFT_U64(pa, 32) | list[j++];
+#endif
+		size = fdt32_to_cpu(list[j++]);
 
 		if (core_is_buffer_inside(region->addr, region->size,
 					  pa, size)) {
-- 
2.47.2

